File: Week 02/id_063/C++/LeetCode_144_063.cpp
#include <stdio.h>
#include <vector>
using namespace std;

struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:

    void dfs(vector<int>& result, TreeNode* p_cur_node) {
        if (p_cur_node == nullptr) {
            return;
        }

        result.push_back(p_cur_node->val);
        dfs(result, p_cur_node->left);
        dfs(result, p_cur_node->right);
    }

    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        dfs(result, root);
        return result;
    }
};

File: Week 02/id_063/Java/LeetCode_144_063.java
/*
思路
简单前序遍历，没什么特殊算法
*/

public class Solution {
    private List<Integer> dfs(List<Integer> result, TreeNode curNode) {
        if (curNode == null) {
            return result;
        }

        result.add(curNode.val);
        return dfs(dfs(result, curNode.left), curNode.right);
    }

    public List<Integer> preorderTraversal(TreeNode root) {
        return dfs(new LinkedList<>(), root);
    }
}